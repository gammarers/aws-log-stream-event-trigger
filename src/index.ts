import { Duration, RemovalPolicy, Stack } from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as destinations from 'aws-cdk-lib/aws-lambda-destinations';
import * as logs from 'aws-cdk-lib/aws-logs';
import { Construct } from 'constructs';
import { ParserFunction } from './funcs/parser-function';

export class LogStreamEventTrigger extends Construct {

  public readonly logStreamSubscriptionFilterDestinationFunction: lambda.IFunction;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    // ðŸ‘‡ Get current account & region
    const account = Stack.of(this).account;
    const region = Stack.of(this).region;
    const partition = Stack.of(this).partition;

    // SubscriptionFilterLogParserFunction

    // ðŸ‘‡ parser Lambda Function
    this.logStreamSubscriptionFilterDestinationFunction = new ParserFunction(this, 'ParserFunction', {
      // functionName: undefined,
      architecture: lambda.Architecture.ARM_64,
      timeout: Duration.seconds(10),
      role: new iam.Role(this, 'ParserFunctionRole', {
        // roleName: `lambda-log-notification-func-${random}-exec-role`,
        assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
        managedPolicies: [
          iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole'),
        ],
        inlinePolicies: {
          ['put-events-policy']: new iam.PolicyDocument({
            statements: [
              new iam.PolicyStatement({
                effect: iam.Effect.ALLOW,
                actions: ['events:PutEvents'],
                resources: [`arn:${partition}:events:${region}:${account}:event-bus/default`],
              }),
            ],
          }),
        },
      }),
      logGroup: new logs.LogGroup(this, 'ParserFunctionLogGroup', {
        //logGroupName: `/aws/lambda/${functionName}`,
        retention: logs.RetentionDays.THREE_MONTHS,
        removalPolicy: RemovalPolicy.RETAIN_ON_UPDATE_OR_DELETE,
      }),
      logFormat: lambda.LogFormat.JSON,
      systemLogLevel: lambda.SystemLogLevel.INFO,
      applicationLogLevel: lambda.ApplicationLogLevel.INFO,
      onSuccess: new destinations.EventBridgeDestination(),
      // onFailure: new lambdaDestinations.EventBridgeDestination(),
    });

  }
}
