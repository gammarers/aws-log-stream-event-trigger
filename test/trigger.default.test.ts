import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { LogStreamEventTrigger } from '../src';

describe('LogStreamEventTrigger default Testing', () => {

  const app = new App();

  const stack = new Stack(app, 'Stack');

  new LogStreamEventTrigger(stack, 'LogStreamEventTrigger');

  const template = Template.fromStack(stack);

  it('Should have one iam role existing', async () => {
    template.resourceCountIs('AWS::IAM::Role', 1);
  });

  it('Should have one lambda function existing', async () => {
    template.resourceCountIs('AWS::Lambda::Function', 1);
  });

  it('Should have one log group existing', async () => {
    template.resourceCountIs('AWS::Logs::LogGroup', 1);
  });

  it('Should match snapshot', () => {
    expect(template.toJSON()).toMatchSnapshot();
  });
});
