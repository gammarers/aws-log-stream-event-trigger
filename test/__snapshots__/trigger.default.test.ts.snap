// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogStreamEventTrigger default Testing Should match snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "LogStreamEventTriggerParserFunction924F8116": {
      "DependsOn": [
        "LogStreamEventTriggerParserFunctionRoleDefaultPolicy79BB7F36",
        "LogStreamEventTriggerParserFunctionRole03C40C6D",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "HASH.zip",
        },
        "Description": "src/funcs/parser.lambda.ts",
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "LoggingConfig": {
          "ApplicationLogLevel": "INFO",
          "LogFormat": "JSON",
          "LogGroup": {
            "Ref": "LogStreamEventTriggerParserFunctionLogGroupCE8F34C7",
          },
          "SystemLogLevel": "INFO",
        },
        "Role": {
          "Fn::GetAtt": [
            "LogStreamEventTriggerParserFunctionRole03C40C6D",
            "Arn",
          ],
        },
        "Runtime": "nodejs22.x",
        "Timeout": 10,
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogStreamEventTriggerParserFunctionEventInvokeConfig5B81DAB6": {
      "Properties": {
        "DestinationConfig": {
          "OnSuccess": {
            "Destination": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition",
                  },
                  ":events:",
                  {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  ":event-bus/default",
                ],
              ],
            },
          },
        },
        "FunctionName": {
          "Ref": "LogStreamEventTriggerParserFunction924F8116",
        },
        "Qualifier": "$LATEST",
      },
      "Type": "AWS::Lambda::EventInvokeConfig",
    },
    "LogStreamEventTriggerParserFunctionLogGroupCE8F34C7": {
      "DeletionPolicy": "RetainExceptOnCreate",
      "Properties": {
        "RetentionInDays": 90,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "LogStreamEventTriggerParserFunctionRole03C40C6D": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "events:PutEvents",
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":events:",
                        {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":event-bus/default",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "put-events-policy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogStreamEventTriggerParserFunctionRoleDefaultPolicy79BB7F36": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "events:PutEvents",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":events:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":event-bus/default",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogStreamEventTriggerParserFunctionRoleDefaultPolicy79BB7F36",
        "Roles": [
          {
            "Ref": "LogStreamEventTriggerParserFunctionRole03C40C6D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
